lib/Stats.js

```diff
@@ -9,29 +9,52 @@ const SizeFormatHelpers = require("./SizeFormatHelpers");
 const formatLocation = require("./formatLocation");
 const identifierUtils = require("./util/identifier");
 
-const optionsOrFallback = (...args) => {
-	let optionValues = [];
-	optionValues.push(...args);
+/**
+ * @typedef {import("./Compilation")} Compilation
+ * @typedef {string | RegExp | ((warning: any) => any)} WarningFilterAtom
+ * @typedef {WarningFilterAtom | WarningFilterAtom[]} WarningFilter
+ */
+
+const optionsOrFallback = (...optionValues) => {
 	return optionValues.find(optionValue => typeof optionValue !== "undefined");
 };
 
 class Stats {
+	/**
+	 * @param {Compilation} compilation
+	 */
 	constructor(compilation) {
+		/** @type {Compilation} */
 		this.compilation = compilation;
+		/** @type {string} */
 		this.hash = compilation.hash;
+		/** @type {undefined | number} */
 		this.startTime = undefined;
+		/** @type {undefined | number} */
 		this.endTime = undefined;
 	}
 
+	/**
+	 * @param {string[]} warnings
+	 * @param {WarningFilter} warningsFilter
+	 * @return {string[]}
+	 */
 	static filterWarnings(warnings, warningsFilter) {
 		// we dont have anything to filter so all warnings can be shown
 		if (!warningsFilter) {
 			return warnings;
 		}
 
 		// create a chain of filters
-		// if they return "true" a warning should be suppressed
-		const normalizedWarningsFilters = [].concat(warningsFilter).map(filter => {
+		let warningsFilters;
```

Comment:
Does this pass compilation? It’s the same name as the parameter, right? 
Seems like a better solution to just overwrite warningsFilter’s value and give the parameter the type WarningFilter | WarningFilter[].

Reply:
No it's not. There is an `s` at the end to denote the array nature of the binding.

Reply:
`const warningsFilters = Array.isArray(warningsFilter) ? warningsFilter : [warningsFilter];` has the same result as this `if` block I think.